### ISSUE TITLE
Double-escaping bug when rendering rich text blocks with Jinja2

### ISSUE THREAD

**Issue by aaugustin** [torchbox/wagtail#2542]
This is similar but distinct from #2541. The combination of these two bugs currently produces triple-escaping :-)

----

You should be able to reproduce this bug with:

- this page model:

```python
from wagtail.wagtailcore import blocks, fields, models

class PageWithRichTextBlocks(models.Page):

    content = fields.StreamField([
        ('stuff', blocks.RichTextBlock(template='stuff.html')),
    ])
```

- this `page_with_blocks.html` template:

```
{{ page.content }}
```

- this `stuff.html` template (an actual example would contain HTML around the value):

```
{{ value }}
```

The HTML source of the rich text shows up in the rendered web page.

----

Taking a clue from `RichTextField`, I tried to apply the `|richtext` filter as follows:

```
{{ value|richtext }}
```

If I were rendering a `RichTextField`, the `|richtext` filter would call `expand_db_html()`, wrap the result in a div and call `mark_safe` on the result, thus returning a safe string.

With a `RichTextBlock`, `value` is already a `RichText` instance and the `|richtext` filter is a no-op. So `RichText.__str__` is called. It returns a safe string. This achieves the desired effect with the Django template language but not with Jinja2, for reasons I didn't investigate fully.

Anyway, the fix for this problem is the following:

```diff
diff --git a/wagtail/wagtailcore/rich_text.py b/wagtail/wagtailcore/rich_text.py
index af0b778..f66f640 100644
--- a/wagtail/wagtailcore/rich_text.py
+++ b/wagtail/wagtailcore/rich_text.py
@@ -4,7 +4,7 @@ import re  # parsing HTML with regexes LIKE A BOSS.
 
 from django.utils.encoding import python_2_unicode_compatible
 from django.utils.html import escape
-from django.utils.safestring import mark_safe
+from django.utils.safestring import SafeText
 
 from wagtail.wagtailcore import hooks
 from wagtail.wagtailcore.models import Page
@@ -180,7 +180,7 @@ def expand_db_html(html, for_editor=False):
 
 
 @python_2_unicode_compatible
-class RichText(object):
+class RichText(SafeText):
     """
     A custom object used to represent a renderable rich text value.
     Provides a 'source' property to access the original source code,
@@ -191,4 +191,4 @@ class RichText(object):
         self.source = (source or '')
 
     def __str__(self):
-        return mark_safe('<div class="rich-text">' + expand_db_html(self.source) + '</div>')
+        return '<div class="rich-text">' + expand_db_html(self.source) + '</div>'
```

(says the guy who made Django's and Jinja2's idea of what a safe string is almost compatible, but not fully, as this bug demonstrates...)

**Comment by gasman** [comment]
Hi @aaugustin - thanks for the report!

Unfortunately your suggested fix doesn't seem to work - inheriting from `SafeText` means that the call to `RichText.__str__()` is skipped, on both Jinja2 and Django templates. (This is shown by the failing test `test_rich_text_is_safe` in #2859.)

**Comment by gasman** [comment]
I'm now pretty sure that this is a bug in [markupsafe](https://github.com/pallets/markupsafe). A minimal Django example that reproduces the problem:

```python
from django.template import engines
from django.utils.safestring import mark_safe

class Pony(object):
    def __str__(self):
        return mark_safe("<h1>PONY</h1>")

pony = Pony()

django_template = engines['django'].from_string('{{ value }}')
django_template.render({'value': pony})
# '<h1>PONY</h1>'

jinja2_template = engines['jinja2'].from_string('{{ value }}')
jinja2_template.render({'value': pony})
# '&lt;h1&gt;PONY&lt;/h1&gt;'
```

Digging into the code generated by Jinja2, it ultimately calls `escape(value)`, where `escape` is provided by markupsafe as a pure-python implementation (which works) and a C one (which doesn't):

```python
from markupsafe import Markup
class Pony(object):
    def __str__(self):
        return Markup("<h1>PONY</h1>")

pony = Pony()

from markupsafe._native import escape
escape(pony)
# Markup('<h1>PONY</h1>')

from markupsafe._speedups import escape
escape(pony)
# Markup('&lt;h1&gt;PONY&lt;/h1&gt;')
```

I think that's as far down the rabbit hole as I want to go today :-) However, that has clued me into a suitable workaround - namely to provide a `__html__` method on RichText that bypasses the whole 'cast this value to a text type' mechanism. This is now implemented on #2859.

**Comment by aaugustin** [comment]
This may also be an interoperability problem between Django and Jinja2...

Indeed, in my experience, defining `__html__` on objects that need to be rendered without escaping works best.

I'll take a deeper look at this issue when I get the opportunity.
